js/ai/src/resource.ts

```diff
@@ -96,24 +99,74 @@ export function defineResource(
   opts: ResourceOptions,
   fn: ResourceFn
 ): ResourceAction {
+  const action = dynamicResource(opts, fn).attach(registry) as ResourceAction;
+  // when defined via defineResource it's not dynamic.
+  delete action.__action.metadata?.dynamic;
+  action.matches = createMatcher(opts.uri, opts.template);
+  registry.registerAction('resource', action);
+  return action;
+}
+
+/**
+ * A dynamic action with a `resource` type. Dynamic resources are detached actions -- not associated with any registry.
+ */
+export type DynamicResourceAction = DetachedAction<
+  typeof ResourceInputSchema,
+  typeof ResourceOutputSchema
+> & {
+  __action: {
+    metadata: {
+      type: 'resource';
+    };
+  };
+  matches(input: ResourceInput): boolean;
+};
+
+/**
+ * Finds a matching resource in the registry. If not found returns undefined.
+ */
+export async function findMatchingResource(
+  registry: Registry,
+  input: ResourceInput
+): Promise<ResourceAction | undefined> {
+  for (const actKeys of Object.keys(await registry.listResolvableActions())) {
+    if (actKeys.startsWith('/resource/')) {
+      const resource = (await registry.lookupAction(actKeys)) as ResourceAction;
+      if (resource.matches(input)) {
+        return resource;
+      }
+    }
+  }
+  return undefined;
+}
+
+/** Checks whether provided object is a dynamic resource. */
+export function isDynamicResource(t: unknown): t is DynamicResourceAction {
```

Comment:
The function name and the type name should match in a type guard, so either `function isDynamicResourceAction ...` or rename the type to DynamicResource ?

Reply:
renamed the function